nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/PutMongoRecord.java

```diff
@@ -0,0 +1,139 @@
+    static final Relationship REL_SUCCESS = new Relationship.Builder().name("success")
+            .description("All FlowFiles that are written to MongoDB are routed to this relationship").build();
+    static final Relationship REL_FAILURE = new Relationship.Builder().name("failure")
+            .description("All FlowFiles that cannot be written to MongoDB are routed to this relationship").build();
+
+    static final PropertyDescriptor RECORD_READER_FACTORY = new PropertyDescriptor.Builder()
+            .name("record-reader")
+            .displayName("Record Reader")
+            .description("Specifies the Controller Service to use for parsing incoming data and determining the data's schema")
+            .identifiesControllerService(RecordReaderFactory.class)
+            .required(true)
+            .build();
+
+    private final static Set<Relationship> relationships;
+    private final static List<PropertyDescriptor> propertyDescriptors;
+
+    static {
+        List<PropertyDescriptor> _propertyDescriptors = new ArrayList<>();
+        _propertyDescriptors.addAll(descriptors);
+        _propertyDescriptors.add(WRITE_CONCERN);
+        _propertyDescriptors.add(RECORD_READER_FACTORY);
+        propertyDescriptors = Collections.unmodifiableList(_propertyDescriptors);
+
+        final Set<Relationship> _relationships = new HashSet<>();
+        _relationships.add(REL_SUCCESS);
+        _relationships.add(REL_FAILURE);
+        relationships = Collections.unmodifiableSet(_relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return propertyDescriptors;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final RecordReaderFactory recordParserFactory = context.getProperty(RECORD_READER_FACTORY)
+                .asControllerService(RecordReaderFactory.class);
+
+        List<Document> inserts = new ArrayList<>();
+        try {
+            RecordReader reader = recordParserFactory.createRecordReader(flowFile, session.read(flowFile), getLogger());
+            RecordSchema schema = reader.getSchema();
+            Record record = null;
+            while ((record = reader.nextRecord()) != null) {
+                Document document = new Document();
+                for (String name : schema.getFieldNames()) {
+                    document.put(name, record.getValue(name));
+                }
+                inserts.add(document);
+            }
+            reader.close();
+        } catch (SchemaNotFoundException | IOException | MalformedRecordException e) {
+            getLogger().error("PutMongoRecord failed to transform records.", e);
```

Comment:
If this happens, we are unable to properly read the incoming data. As-is, it would send potentially many records and then route to success. We should instead be routing the FlowFile to the 'failure' relationship and returning at this point.